java/feathub-connectors/flink-connectors/flink-connector-prometheus/src/main/java/com/alibaba/feathub/flink/connectors/PrometheusSinkFunction.java

```diff
@@ -0,0 +1,353 @@
+import static com.alibaba.feathub.flink.connectors.PrometheusConfigs.EXTRA_LABELS;
+import static com.alibaba.feathub.flink.connectors.PrometheusConfigs.HOST_URL;
+import static com.alibaba.feathub.flink.connectors.PrometheusConfigs.JOB_NAME;
+
+/**
+ * A {@link org.apache.flink.streaming.api.functions.sink.SinkFunction} that writes to a Prometheus
+ * PushGateway. Each column can be configured with a json map via the column comment.
+ *
+ * <p>Note that the SinkFunction only support writing numeric columns or map columns map from string
+ * to numeric values.
+ *
+ * <p>The following examples illustrate how to define a Prometheus Sink table:
+ *
+ * <pre>
+ *     TableDescriptor.forConnector("prometheus")
+ *         .schema(
+ *                 Schema.newBuilder()
+ *                         .column(
+ *                                 "metric_map",
+ *                                 DataTypes.MAP(DataTypes.STRING(), DataTypes.INT()))
+ *                         .withComment("{\"key_label_name\":\"count_map_key\"}")
+ *                         .column("metric", DataTypes.DOUBLE())
+ *                         .withComment("{\"metric_name\":\"metric\"}")
+ *                         .build())
+ *         .option("hostUrl", "localhost:9091")
+ *         .option("jobName", "exampleJob")
+ *         .option("extraLabels", "k1=v1;k2=v2")
+ *         .build();
+ * </pre>
+ */
+public class PrometheusSinkFunction extends RichSinkFunction<RowData> {
+    private static final Logger LOG = LoggerFactory.getLogger(PrometheusSinkFunction.class);
+
+    private final ReadableConfig config;
+    private CollectorRegistry registry;
+    private PushGateway pushGateway;
+    private String jobName;
+    private Map<String, String> extraLabels;
+    private final List<DataType> fieldTypes;
+    private final List<MetricMeta<Gauge.Child>> metricMetas;
+    private final Map<String, SimpleCollector<Gauge.Child>> nameToCollectors;
+
+    public PrometheusSinkFunction(ReadableConfig config, ResolvedSchema schema) {
+        this.config = config;
+        this.fieldTypes = schema.getColumnDataTypes();
+        this.nameToCollectors = new HashMap<>();
+
+        final List<Column> columns = schema.getColumns();
+        this.metricMetas = new ArrayList<>(columns.size());
+        for (Column column : columns) {
+            metricMetas.add(MetricMeta.fromColumn(column));
+        }
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        super.open(parameters);
+
+        registry = new CollectorRegistry();
+        pushGateway = new PushGateway(config.get(HOST_URL));
+        jobName = config.get(JOB_NAME);
+        extraLabels = parseGroupingKey(config.get(EXTRA_LABELS));
+
+        for (int i = 0; i < metricMetas.size(); i++) {
+            final LogicalType logicalType = fieldTypes.get(i).getLogicalType();
+            verifyType(logicalType);
```

Comment:
Let's also verify that metrics with the same name should have the same label names. It seems that user would get an exception that says "the length of label names and labels mismatch" when the above rule is violated in the current implementation, which is not straight forward enough to know how to correct the mistake.